```js
package test1;

public class Test {
    public static void main(String[] args) {
        Thread runThread = Thread.currentThread();
        System.out.println(runThread.getName() + " " + runThread.getId());
        Thread t1 = new Thread();
        System.out.println(t1.getName() + " " + t1.getId());
        Thread t2 = new Thread();
        System.out.println(t2.getName() + " " + t2.getId());
        Thread t3 = new Thread();
        System.out.println(t3.getName() + " " + t3.getId());
        Thread t4 = new Thread();
        System.out.println(t4.getName() + " " + t4.getId());
        Thread t5 = new Thread();
        System.out.println(t5.getName() + " " + t5.getId());
    }
}
```

![image](https://github.com/user-attachments/assets/1f903c8c-b678-44f2-98c4-43efe6d3602a)


程序运行结果如图1-36所示。

从运行结果来看，当前执行代码的线程名为main，线程id值为1。

而Thread-0线程的id值直接到达11，说明中间有9个id值被隐藏的线程占有。

# 1.11 停止线程

使用Java内置支持多线程的Thread类去设计多线程应用是很常见的事情，然而多线程也给开发人员带来了一些新的挑战，如果处理不好就会导致超出预期的行为及难以定位的错误。

停止线程是多线程开发的一个重要的技术点，掌握此技术可以对线程的停止进行有效的处理，停止线程在Java语言中并不像break语句那样干脆，需要一些技巧性的处理。

本节将讨论如何更好地停止一个线程。停止一个线程意味着在线程处理完任务之前停止正在做的操作，也就是放弃当前的操作，虽然这看起来非常简单，但是必须要做好防范措施，以便达到预期的效果。

在停止一个线程可以使用Thread.stop()方法，但并不推荐使用，该方法确实可以停止一个正在运行的线程，但已经被弃用作废，即在将来的Java版本中，这个方法将不可用或不被支持。

大多数情况下，使用Thread.interrupt()方法停止一个线程，但这个方法不会终止一个正在运行的线程，还需要加入一个判断才可以完成线程的停止。关于此知识点，后面有专门的章节进行介绍，这里不再赘述。

在Java中有三种方法可以使正在运行的线程终止运行。

1）使用退出标志使线程正常退出。 


2）使用stop()方法强行终止线程，但是这个方法不推荐使用，因为stop()和suspend()和resume()一样，都是作废过期的方法，使用它们可能发生不可预料的结果。

3）使用interrupt()方法中断线程。

这三种方法都会在后面的章节进行介绍。

# 1.11.1 停止不了的线程

本示例将调用interrupt()方法来停止线程，但interrupt()方法的使用效果并不像for+break语句那样，可以马上停止循环，该方法仅仅是在当前线程中打了一个停止的标记，并不是真正的停止线程。

创建名称为t11的项目，文件MyThread.java代码如下：

```java
public class MyThread extends Thread {
    @Override
    public void run() {
        super.run();
        for (int i = 0; i < 500000; i++) {
            System.out.println("i=" + (i + 1));
        }
    }
}
```

运行类Run.java代码如下：

```java
package test;

import exthread.MyThread;

public class Run {
    public static void main(String[] args) throws InterruptedException {
        MyThread thread = new MyThread();
        thread.start();
        Thread.sleep(2000);
        thread.interrupt();
        System.out.println("zzzzzzzz");
    }
}
```

程序运行结果如图1-37所示。

把IDE控制台中的日志复制到文本编辑器软件中，显示的行数是500001行，如图1-38所示。

在331063行处输出了zzzzzzzz，说明sleep时间为2秒时，for语句跑了331063次循环，日志如下：
```
i=331062
i=331063
zzzzzzzz
```
图1-37 正常循环打印50万次 

![image](https://github.com/user-attachments/assets/7aea9e0c-7e76-4660-8fd8-0d37a86a9279)

i=331064

i=331065

从运行结果来看，调用interrupt方法并没有将线程停止，那如何停止线程呢？

![image](https://github.com/user-attachments/assets/0d986366-cc16-4802-8eaf-fb1de99448ce)


# 1.11.2 判断线程是不是停止状态

在介绍如何停止线程之前，先来看一下如何判断线程的状态已经是停止的。在Java的SDK中，Thread.java类里提供了两种判断方法。

1）public static boolean interrupted()：测试currentThread()是否已经中断。

2）public boolean this.isInterrupted()：测试this关键字所在线程类的对象是否已经中断。

其中，interrupted()方法的声明如图1-39所示。

isInterrupted()方法的声明如图1-40所示。

![image](https://github.com/user-attachments/assets/00f194a7-79f1-4853-a281-55e3724cd5a0)


interrupted

public static boolean interrupted()

图1-39 interrupted方法的声明

isInterrupted

public boolean isInterrupted()

图1-40 isInterrupted方法的声明

这两种方法有什么区别呢？先来看看this.isInterrupted()方法的解释：测试当前线程是否已经中断。当前线程是指执行this.isInterrupted()方法的线程，为了对此方法有更深入的了解，创建项目t12，类MyThread.java代码如下：
```java
public class MyThread extends Thread {
    @Override
    public void run() {
        super.run();
        for (int i = 0; i < 500000; i++) {
            System.out.println("i=" + (i + 1));
        }
    }
}
```
类Run.java代码如下：
```java
public class Run {
    public static void main(String[] args) {
        try {
            MyThread thread = new MyThread();
            thread.start();
            Thread.sleep(1000);
            thread.interrupt();
            //Thread.currentThread().interrupt();
            System.out.println("是否停止1？="+thread.isInterrupted());
            System.out.println("是否停止2？="+thread.interrupted());
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}
```
图1-38 确认是500001行并且线程未停止 

```js
    System.out.println("main catch");
    e.printStackTrace();
}
System.out.println("end!");
}
```
![image](https://github.com/user-attachments/assets/70d70a1c-160a-4ae7-b7b7-8cf6f8f04eae)


程序运行结果如图1-41所示。

类Run.java中虽然是在thread对象上调用代码：

thread.interrupt();

来停止thread对象所代表的线程，在后面又使用代码：

System.out.println("是否停止1？="+thread.interrupted());

System.out.println("是否停止2？="+thread.interrupted());

来判断thread对象所代表的线程是否停止，但从控制台输出的结果来看，线程并未停止，证明了interrupted()方法的解释：测试当前线程是否已经中断。这个当前线程是main，从未中断过，所以输出的结果是两个false。

注 意

测试代码中使用thread.interrupted()来判断currentThread()是否被中断，也可以使用Thread.interrupted()进行判断，因为在Thread.java类中调用静态static方法时，大多都是针对currentThread()线程进行操作的。


如何使main线程有中断效果呢？创建Run2.java代码如下：

```java
public class Run2 {
    public static void main(String[] args) {
        Thread.currentThread().interrupt();
        System.out.println("是否停止1？=" + Thread.interrupted());
        System.out.println("是否停止2？=" + Thread.interrupted());
        System.out.println("end!");
    }
}
```
程序运行结果如图1-42所示。

![image](https://github.com/user-attachments/assets/a1af3d59-fea3-4544-af82-888f565627cf)


从上述结果来看，interrupted()方法的确判断了当前线程是否是停止状态，但为什么第二个Boolean布尔值是false呢？interrupted方法在官方帮助文档中的解释如下：

测试当前线程是否已经中断。线程的中断状态由该方法清除。换句话说，如果连续两次调用该方法，则第二次调用将返回false（在第一次调用已清除其中断状态之后，且第二次调用检验完中断状态前，当前线程再次被中断的情况除外）。

图1-41 运行结果

图1-42 主线程main已是停止状态 


文档已经解释得很详细，interrupted()方法具有清除状态的功能，所以第二次调用interrupted()方法返回的值是false。

介绍完interrupted()方法后再来看一下isInterrupted()方法，声明如下：

public boolean isInterrupted()

从声明中可以看出isInterrupted()方法不是静态的，具体取决于调用这个方法的线程对象。

继续创建Run3.java类，代码如下：

```java
public class Run3 {
    public static void main(String[] args) {
        try {
            MyThread thread = new MyThread();
            thread.start();
            Thread.sleep(1000);
            thread.interrupt();
            System.out.println("是否停止1？="+thread.isInterrupted());
            System.out.println("是否停止2？="+thread.isInterrupted());
        } catch (InterruptedException e) {
            System.out.println("main catch");
            e.printStackTrace();
        }
        System.out.println("end!");
    }
}
```
程序运行结果如图1-43所示。

![image](https://github.com/user-attachments/assets/b7465bad-9da1-4875-ab94-1653a527edc4)


从结果可以看到，isInterrupted()方法并未清除状态标志，不具有此功能，所以输出两个true。但是也有非常非常小的概率先输出一个false，如中断的标记还未更新。

最后，再来看一下这两种方法的解释。

1）this.interrupted()：测试当前线程是否已经是中断状态，执行后具有清除状态标志值为false的功能。

2）this.isInterrupted()：测试所在Thread线程对象是否已经是中断状态，不清除状态标志。

# 1.11.3 清除中断状态的使用场景

this.interrupted()方法具有清除状态标志值为false的功能，借用此特性可以实现一些效果，这里简单测试一下。

本节要实现的效果是在MyThread线程中向list1和list2存放数据，基于职责单一原则，MyThread线程只负责存放数据，不负责处理存放的数据量，数据量由main线程进行处理。创建测试项目clearStatus。

创建工具类代码如下：

图1-43 已经是停止状态 

### 工具类代码（package tools）
```java
import java.util.ArrayList;
public class Box {
    public static ArrayList list1 = new ArrayList();
    public static ArrayList list2 = new ArrayList();
}
```
### 线程类代码（package test1）
```java
import tools.Box;
public class MyThread extends Thread {
    @Override
    public void run() {
        try {
            while (true) {
                if (this.isInterrupted()) {
                    throw new InterruptedException("线程被中断！");
                }
                //模拟执行任务的耗时，不能使用sleep，遇到interrupt()方法会出现异常，所以用for循环实现
                for (int i = 0; i < 10000; i++) {
                    new String("" + Math.random());
                }
                Box.list1.add("生产数据A");
                System.out.println("list1 size=" + Box.list1.size());
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        try {
            while (true) {
                if (this.isInterrupted()) {
                    throw new InterruptedException("线程被中断！");
                }
                for (int i = 0; i < 10000; i++) {
                    new String("" + Math.random());
                }
                Box.list2.add("生产数据B");
                System.out.println("list2 size=" + Box.list2.size());
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}
```
### 运行类代码（package test1）
```java
import tools.Box;
public class Test1 {
    public static void main(String[] args) throws InterruptedException {
        MyThread t = new MyThread();
        t.start();
        boolean list1Isinterrupted = false;
        boolean list2Isinterrupted = false;
        while (t.isAlive()) {
            if (Box.list1.size() > 500 && list1Isinterrupted == false) {
                t.interrupt();
                list1Isinterrupted = true;
            }
            if (Box.list2.size() > 600 && list2Isinterrupted == false) {
                t.interrupt();
                list2Isinterrupted = true;
            }
            Thread.sleep(50);
        }
    }
}
```
运行程序后控制台输出结果：
```
list1 size=511
list1 size=512
java.lang.InterruptedException: 线程被中断！
at test1.MyThread.run(MyThread.java:11)
java.lang.InterruptedException: 线程被中断！
at test1.MyThread.run(MyThread.java:28)
```

从输出结果可以发现，并未向list2中添加数据，因为使用了isInterruptedO方法作为判断条件，该方法不会清除中断状态，多次调用isInterrupted()方法的返回值永远是true。同时，进程状态一直呈红色，说明main线程一直在while(true)循环执行，程序出现了世误。

继续实验。

创建线程类代码如下:

### 改进后的线程类代码（package test2）
```java
import tools.Box;
public class MyThread extends Thread {
    @Override
    public void run() {
        try {
            while (true) {
                if (this.interrupted()) {
                    throw new InterruptedException("线程被中断！");
                }
                //模拟执行任务的耗时，不能使用sleep，遇到interrupt()方法会出现异常，所以用for循环实现
                for (int i = 0; i < 10000; i++) {
                    new String("" + Math.random());
                }
                Box.list1.add("生产数据A");
                System.out.println("list1 size=" + Box.list1.size());
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        try {
            while (true) {
                if (this.interrupted()) {
                    throw new InterruptedException("线程被中断！");
                }
                for (int i = 0; i < 10000; i++) {
                    new String("" + Math.random());
                }
                Box.list2.add("生产数据B");
                System.out.println("list2 size=" + Box.list2.size());
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}
```
### 改进后的运行类代码（package test2）
```java
import tools.Box;
public class Test2 {
    public static void main(String[] args) throws InterruptedException {
        MyThread t = new MyThread();
        t.start();
        boolean list1Isinterrupted = false;
        boolean list2Isinterrupted = false;
        while (t.isAlive()) {
            if (Box.list1.size() > 500 && list1Isinterrupted == false) {
                t.interrupt();
                list1Isinterrupted = true;
            }
            if (Box.list2.size() > 600 && list2Isinterrupted == false) {
                t.interrupt();
                list2Isinterrupted = true;
            }
            Thread.sleep(50);
        }
    }
}
```
程序运行结果：
```
list1 size=1
list1 size=2
......
list1 size=503
list1 size=504
list1 size=505
list1 size=506
list1 size=507
list1 size=508
java.lang.InterruptedException: 线程被中断！
at test2.MyThread.run(MyThread.java:11)
list2 size=1
list2 size=2
......
list2 size=616
list2 size=617
list2 size=618
list2 size=619
java.lang.InterruptedException: 线程被中断！
at test2.MyThread.run(MyThread.java:28)
```

成功向2个list中添加数据，并且进程按钮 为灰色，表示进程销毁，程序没有出现问题。

### 1.11.4 能停止的线程——异常法
根据前面学过的知识点，只需要通过线程中的for语句来判断线程是否是停止状态即可判断后面的代码是否可运行，如果是停止状态，则后面的代码不再运行即可。

创建实验项目t13，类MyThread.java代码如下：
```java
public class MyThread extends Thread {
    @Override
    public void run() {
        super.run();
        for (int i = 0; i < 500000; i++) {
            if (this.interrupted()) {
                System.out.println("已经是停止状态了！我要退出了！");
                break;
            }
            System.out.println("i=" + (i + 1));
        }
    }
}
```
类Run.java代码如下：

